### PR TITLE
Ensure Newly Added Instances Have Up-to-Date Index

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
As of today, STOW is performed in multiple steps:
1. Validate instance
2. Create instance entry in SQL with index
3. Upload DCM file and metadata to blob storage
4. Finalize instance status

However, this creates a problem for re-indexing when a new tag is added sometime between steps (2) and (4) above. In this scenario, an instance's index has been added, but it is out-of-date. Unfortunately, until we're guaranteed that the blob entry exists, there is no way for the re-indexing process to update the entry. So the real solution is to update the index on STOW has the last step.

To facilitate this change, this PR includes the following:
- New stored procedures `BeginAddInstance` and `EndAddInstance` which serve to bookend the STOW operation where `EndAddInstance` includes the extended query tags
- Rename the methods in `IIndexStore` to `BeginCreateInstanceIndexAsync` and `EndCreateInstanceIndexAsync`
- Refactor `StoreOrchestrator` to retry the `EndCreateInstanceIndexAsync` operation should it fail due to invalid query tags
- Re-use caching logic in `VersionCache<T>` by creating new `AsyncCache<T>`
- Change `QueryTagService` API to allow users to force the refresh of the query tags
- Remove `TagVersion` from `ExtendedQueryTag` table

## Related issues
[AB#81803](https://microsofthealth.visualstudio.com/Health/_workitems/edit/81803)

## Testing
Test pipelines

### CODE REVIEW

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV4.cs [hunk 1]
```diff
@@ -96,6 +79,52 @@ public override async Task<long> CreateInstanceIndexAsync(DicomDataset instance,
...
+            using (SqlConnectionWrapper sqlConnectionWrapper = await SqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                VLatest.EndAddInstance.PopulateCommand(
+                    sqlCommandWrapper,
+                    instance.GetSingleValueOrDefault(DicomTag.StudyInstanceUID, string.Empty),
+                    instance.GetSingleValueOrDefault(DicomTag.SeriesInstanceUID, string.Empty),
+                    instance.GetSingleValueOrDefault(DicomTag.SOPInstanceUID, string.Empty),
+                    watermark,
+                    UlongToRowVersion(rows.MaxVersion),
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV4.cs:109]: I'm starting to wonder if we really need this field or whether we can just use `TagKey`?

File: src/Microsoft.Health.Dicom.Core/Features/Store/IStoreOrchestrator.cs [hunk 2]
```diff
@@ -14,6 +15,11 @@ namespace Microsoft.Health.Dicom.Core.Features.Store
     /// </summary>
     public interface IStoreOrchestrator
     {
+        /// <summary>
+        /// Occurs when the locally resolved set of query tags is missing at least one newly added entry in the store.
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Store/IStoreOrchestrator.cs:19]: Oh true!

File: src/Microsoft.Health.Dicom.Core/Features/Store/IIndexDataStore.cs [hunk 3]
```diff
@@ -69,13 +68,15 @@ public interface IIndexDataStore
...
-        /// <param name="versionedInstanceIdentifier">The DICOM instance identifier.</param>
-        /// <param name="status">The status to update to.</param>
+        /// <param name="dicomDataset">The DICOM dataset whose status should be updated.</param>
+        /// <param name="watermark">The DICOM instance watermark.</param>
+        /// <param name="queryTags">Queryable dicom tags</param>
+        /// <param name="allowExpiredTags">Optionally allow an out-of-date snapshot of <paramref name="queryTags"/>.</param>
         /// <param name="cancellationToken">The cancellation token.</param>
         /// <returns>A task that represents the asynchronous update operation.</returns>
-        Task UpdateInstanceIndexStatusAsync(VersionedInstanceIdentifier versionedInstanceIdentifier, IndexStatus status, CancellationToken cancellationToken = default);
+        Task EndCreateInstanceIndexAsync(DicomDataset dicomDataset, long watermark, IEnumerable<QueryTag> queryTags, bool allowExpiredTags = false, CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Store/IIndexDataStore.cs:79]: Only in tests

File: src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/ExtendedQueryTagDataRows.cs [hunk 4]
```diff
@@ -13,10 +13,16 @@ namespace Microsoft.Health.Dicom.SqlServer.Features.ExtendedQueryTag
     /// </summary>
     internal class ExtendedQueryTagDataRows
     {
+        public int MaxTagKey { get; set; }
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/ExtendedQueryTagDataRows.cs:16]: I defaulted no rows to `0` in the stored procedure

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.sql [hunk 5]
```diff
@@ -736,14 +734,10 @@ GO
...
 --         * PersonName extended query tag data
---     @initialStatus
---         * Initial status
---     @maxTagVersion
---         * Max ExtendedQueryTag version
 -- RETURN VALUE
 --     The watermark (version).
 ------------------------------------------------------------------------
-CREATE PROCEDURE dbo.AddInstance
+CREATE OR ALTER PROCEDURE dbo.AddInstance
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.sql:740]: We'll keep it for backwards compatibility

File: src/Microsoft.Health.Dicom.Core/Features/Store/IStoreOrchestrator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Store/IStoreOrchestrator.cs:19]: I've changed the logic to consistently throw an exception only when we're potentially missing a tag. If the latest tag was deleted, we should be ok.

**Review by pengchen0692 — APPROVED** [review]

**Pull request merged**